src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,87 +87,102 @@
-          (.clear new-pool)
-          (.insertPill new-pool (PoisonPill. e))
+          (.clear pool)
+          (jruby-internal/insert-poison-pill pool e)
           (throw (IllegalStateException.
-                  "There was a problem adding a JRubyInstance to the pool."
-                  e)))))
-    ;; Add a "RetryPoisonPill" to the pool in case something else is in the
-    ;; process of borrowing from the old pool.
-    (.insertPill old-pool (RetryPoisonPill. old-pool))))
+                  "There was a problem creating a JRubyInstance for the pool."
+                  e))))))
+  (if refill?
+    (log/info "Finished draining and refilling pool.")
+    (log/info "Finished draining pool.")))
+
+(schema/defn ^:always-validate
+  drain-and-refill-pool!
+  "Borrow and destroy all the jruby instances, optionally refilling the
+  pool with fresh jrubies. Locks the pool in order to drain it, but releases
+  the lock before destroying the instances and refilling the pool"
+  [pool-context :- jruby-schemas/PoolContext
+   pool-state :- jruby-schemas/PoolState
+   refill? :- schema/Bool]
+  (if refill?
+    (log/info "Draining and refilling JRuby pool.")
+    (log/info "Draining JRuby pool."))
+  (let [shutdown-on-error (get-shutdown-on-error-fn pool-context)
+        old-instances (shutdown-on-error #(collect-all-jrubies pool-state))]
+    (log/info "Borrowed all JRuby instances, proceeding with cleanup.")
+    (send-agent (get-modify-instance-agent pool-context)
+                #(cleanup-and-refill-pool pool-context pool-state old-instances refill?))))
 
 (schema/defn ^:always-validate
   flush-pool-for-shutdown!
   "Flush of the current JRuby pool when shutting down during a stop.
   Delivers the on-complete promise when the pool has been flushed."
-  ;; Since this function is only called by the pool-agent, we know that if we
+  ;; Since the drain-pool! function takes the pool lock, we know that if we
   ;; receive multiple flush requests before the first one finishes, they will
-  ;; be queued up and we don't need to worry about race conditions between the
-  ;; steps we perform here in the body.
-  [pool-context :- jruby-schemas/PoolContext
-   on-complete :- IDeref]
-  (try
-    (log/info "Flush request received; creating new JRuby pool.")
-    (let [{:keys [config]} pool-context
-          new-pool-state (jruby-internal/create-pool-from-config config)
-          new-pool (:pool new-pool-state)
-          old-pool-state (jruby-internal/get-pool-state pool-context)
-          old-pool (:pool old-pool-state)
-          old-pool-size (:size old-pool-state)]
-      (when-not (pool-initialized? old-pool-size old-pool)
-        (throw (IllegalStateException. "Attempting to flush a pool that does not appear to have successfully initialized. Aborting.")))
-      (.insertPill new-pool (ShutdownPoisonPill. new-pool))
-      (swap-and-drain-pool! pool-context old-pool-state new-pool-state false))
-    (finally
-      (deliver on-complete true))))
```

Comment:
@camlow325 yeah I think you're right. It would be nice to have a way to tell when the flush has completed. Right now when this function returns you only know that all the instances have been borrowed. Adding the promise back in seems like a good idea. 

The `flush-and-repopulate-pool!` fn can just ignore the promise since it doesn't need to block until the instances have been totally destroyed and recreated